common/dconfig/configmgr.go

```diff
@@ -339,46 +369,75 @@ func (cfgMgr *CassandraConfigManager) mkKVTreeForSvc(service string, items []*m.
 			continue
 		}
 
-		versions[version] = struct{}{}
+		if sku != wildcardToken && host != wildcardToken {
+			key := strings.Join([]string{service,version, sku, host,cfgKey}, ".")
+			sku = wildcardToken
+			items[i].Sku = common.StringPtr(wildcardToken)
+			cfgMgr.logger.WithField(key, cfgValue).Warn("Forcing sku=*;host specific items MUST be sku agnostic")
+		}
+
+		allSkus[sku] = struct{}{}
+		allVersions[version] = struct{}{}
 
 		if version == wildcardToken {
 			if sku == wildcardToken {
-				if host != wildcardToken {
-					// when hostname is present, we always expect the sku
-					// this is an invariant for the hierarchial config store
-					cfgMgr.logger.Errorf("Skipping illegal config item with wildcard sku, item=%v", item)
+				if host == wildcardToken {
+					// service.*.*.*.key, add it at the root level
+					root.keyValues[cfgKey] = cfgValue
 					continue
 				}
-				// service.*.*.*.key, add it at the root level
-				root.keyValues[cfgKey] = cfgValue
+				// 1(a) keep track of svc.* config items
+				wildcardVersionItems = append(wildcardVersionItems, item)
 				continue
 			}
-			// keep track of service.* wildcards separtely
-			// they are the defaults for all service.version.*
-			wildcardVersionItems = append(wildcardVersionItems, item)
+			// 1(b) Keep track of svc.*.sku config items
+			wildcardVersionSkuItems = append(wildcardVersionSkuItems, item)
+			continue
+		}
+
+		if sku == wildcardToken {
+			defaults, ok := versionToWildcardSkuItems[version]
+			if !ok {
+				defaults = make([]*m.ServiceConfigItem, 0, 4)
+				versionToWildcardSkuItems[version] = defaults
+			}
+			// 1(c) keep track of svc.ver.* items
+			versionToWildcardSkuItems[version] = append(defaults, item)
 			continue
 		}
 
 		addToKVTree(root, version, sku, host, cfgKey, cfgValue)
 	}
 
-	// now apply the wildcarded version items for every
-	// version that we know of, including the wildcard
-	for _, item := range wildcardVersionItems {
+	// apply svc.ver.* items as default for every svc.ver.sku
+	for ver, items := range versionToWildcardSkuItems {
+		for _, item := range items {
+			for s := range allSkus {
+				addToKVTree(root, ver, s, item.GetHostname(), item.GetConfigKey(), item.GetConfigValue())
+			}
+		}
+	}
 
-		sku := item.GetSku()
-		host := item.GetHostname()
-		cfgKey := item.GetConfigKey()
-		cfgValue := item.GetConfigValue()
+	// apply svc.*.sku. items as default for every svc.ver.sku items
+	for _, item := range wildcardVersionSkuItems {
+		for ver := range allVersions {
+			addToKVTree(root, ver, item.GetSku(), item.GetHostname(), item.GetConfigKey(), item.GetConfigValue())
+		}
+	}
 
-		for ver := range versions {
-			addToKVTree(root, ver, sku, host, cfgKey, cfgValue)
+	// apply svc.*. items as default for every svc.ver.*
+	for _, item := range wildcardVersionItems {
+ 		for ver := range allVersions {
```

Comment:
cosmetic: there's a 'space' in the beginning of this line (that my editor) points out/fixes.